// flkit/lib/util/error.js
'use strict';

/**
 * error class
 */

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BaseError = function () {
  /**
   * init
   * @param  {String} message [error message]
   * @param  {Number} line    [error line]
   * @param  {Number} col     [error col]
   * @return {Object}         []
   */
  function BaseError() {
    var message = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    var line = arguments[1];
    var col = arguments[2];
    (0, _classCallCheck3.default)(this, BaseError);

    this.message = message;
    if (line !== undefined) {
      this.line = line + 1;
      this.col = col + 1;
    }
    this.stack = new Error(this.toString()).stack;
  }
  /**
   * toString
   * @return {String} get error string
   */


  BaseError.prototype.toString = function toString() {
    if (this.line !== undefined) {
      return this.message + (' (line: ' + this.line + ', col: ' + this.col + ')\n');
    }
    return this.message + '\n';
  };

  return BaseError;
}();

exports.default = BaseError;
