// flkit/lib/template/php.js
'use strict';

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _base = require('./base.js');

var _base2 = _interopRequireDefault(_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PHPTemplate = function (_Base) {
  (0, _inherits3.default)(PHPTemplate, _Base);

  function PHPTemplate() {
    (0, _classCallCheck3.default)(this, PHPTemplate);
    return (0, _possibleConstructorReturn3.default)(this, _Base.apply(this, arguments));
  }

  /**
   * check text has tpl
   * @type {Boolean}
   */
  PHPTemplate.prototype.hasTpl = function hasTpl(text, ld /*, rd */) {
    return text.indexOf(ld) > -1;
  };
  /**
   * get template matched
   */


  PHPTemplate.prototype.getMatched = function getMatched(ld, rd, tokenizeInstance) {
    return this._getMatched(ld, rd, tokenizeInstance, {
      ignoreEnd: true,
      quote: true,
      multiComment: true
    });
  };

  return PHPTemplate;
}(_base2.default);

exports.default = PHPTemplate;
