// flkit/lib/index.js
'use strict';

exports.__esModule = true;
exports.createRawToken = exports.createToken = exports.HtmlCompress = exports.htmlToken2Text = exports.HtmlTokenize = exports.CssCompress = exports.cssToken2Text = exports.CssSelectorTokenize = exports.CssTokenize = exports.Template = exports.TokenType = undefined;
exports.hasTpl = hasTpl;
exports.getHtmlAttrValue = getHtmlAttrValue;
exports.setHtmlAttrValue = setHtmlAttrValue;

var _token_type = require('./util/token_type.js');

var _token_type2 = _interopRequireDefault(_token_type);

var _base = require('./template/base.js');

var _base2 = _interopRequireDefault(_base);

var _tokenize = require('./lang/css/tokenize.js');

var _tokenize2 = _interopRequireDefault(_tokenize);

var _selector_tokenize = require('./lang/css/selector_tokenize.js');

var _selector_tokenize2 = _interopRequireDefault(_selector_tokenize);

var _tokenize3 = require('./lang/html/tokenize.js');

var _tokenize4 = _interopRequireDefault(_tokenize3);

var _util = require('./lang/html/util.js');

var _util2 = require('./lang/css/util.js');

var _compress = require('./lang/html/compress.js');

var _compress2 = _interopRequireDefault(_compress);

var _compress3 = require('./lang/css/compress.js');

var _compress4 = _interopRequireDefault(_compress3);

var _util_ext = require('./util/util_ext.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// import BaseTokenize from './util/tokenize.js';
var TokenType = exports.TokenType = _token_type2.default;
var Template = exports.Template = _base2.default;

var CssTokenize = exports.CssTokenize = _tokenize2.default;
var CssSelectorTokenize = exports.CssSelectorTokenize = _selector_tokenize2.default;
var cssToken2Text = exports.cssToken2Text = _util2.token2Text;
var CssCompress = exports.CssCompress = _compress4.default;

var HtmlTokenize = exports.HtmlTokenize = _tokenize4.default;
var htmlToken2Text = exports.htmlToken2Text = _util.token2Text;
var HtmlCompress = exports.HtmlCompress = _compress2.default;

var createToken = exports.createToken = _util_ext.createToken;
var createRawToken = exports.createRawToken = _util_ext.createRawToken;

/**
 * has template syntax in text
 */
function hasTpl(text) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  var instance = new _base2.default(text, options);
  return instance.hasTpl();
}

/**
 * get html attribute value
 */
function getHtmlAttrValue(attrs, name) {
  var value = void 0;
  attrs.some(function (item) {
    if (item.nameLowerCase === name) {
      value = item.value;
      return true;
    }
  });
  return value;
}
/**
 * set html attribute value
 */
function setHtmlAttrValue(attrs, name, value) {
  var flag = attrs.some(function (item) {
    if (item.nameLowerCase === name) {
      item.value = value;
      return true;
    }
  });
  if (!flag) {
    attrs.push({
      name: name,
      value: value,
      quote: '"',
      nameLowerCase: name
    });
  }
  return attrs;
}
